src/pg_query_summary_truncate.c

```diff
@@ -0,0 +1,673 @@
+
+	if (state.error)
+		return state.error;
+
+	list_sort(state.truncations, cmp_possible_truncation_depth);
+	return apply_truncations(summary, tree, &state, truncate_limit);
+}
+
+static void truncate_mbstr(char *mbstr, size_t max_chars)
+{
+	// Determine the number of characters in mbstr.
+	int n_chars = pg_mbstrlen(mbstr);
+
+	// If we don't need to truncate the string, return immediately.
+	if (n_chars <= max_chars)
+		return;
+
+	// Determine how many bytes hold `max_chars - 3`.
+	int n_bytes = pg_mbcharcliplen(mbstr, n_chars, max_chars - 3);
+
+	// Actually truncate it.
+	strncpy(mbstr + n_bytes, "...", 4);
+	mbstr[n_bytes + 3] = '\0';
+}
+
+static void add_truncation(TruncationState *state, enum TruncationAttr attr,
+		Node *node, int32_t length)
+{
+	PossibleTruncation *truncation = palloc(sizeof(PossibleTruncation));
+
+	truncation->attr = attr;
+	truncation->node = node;
+	truncation->depth = state->depth;
+	truncation->length = length;
+
+	state->truncations = lappend(state->truncations, truncation);
+}
+
+static void add_truncation_where_clause(TruncationState *state, Node *node, Node *whereClause)
+{
+	if (whereClause == NULL)
+		return;
+
+	add_truncation(state,
+			TRUNCATION_WHERE_CLAUSE,
+			node,
+			where_clause_len(whereClause));
+}
+
+static bool
+summary_truncation_options(Node *node, TruncationState *state)
+{
+	if (node == NULL)
+		return false;
+
+	switch (nodeTag(node))
+	{
+		case T_RawStmt:
+			// For some reason our statement has been trapped in a RawStmt.
+			// Free them.
+			return summary_truncation_options(castNode(RawStmt, node)->stmt, state);
+
+		case T_SelectStmt:
+			{
+				SelectStmt *stmt = castNode(SelectStmt, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							select_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				if (stmt->valuesLists != NULL)
+					add_truncation(state,
+							TRUNCATION_VALUES_LISTS,
+							node,
+							select_values_lists_len(stmt->valuesLists));
+
+				break;
+			}
+
+		case T_UpdateStmt:
+			{
+				UpdateStmt *stmt = castNode(UpdateStmt, node);
+
+				if (stmt->targetList != NULL)
+					add_truncation(state,
+							TRUNCATION_TARGET_LIST,
+							node,
+							update_target_list_len(stmt->targetList));
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_DeleteStmt:
+			{
+				DeleteStmt *stmt = castNode(DeleteStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_CopyStmt:
+			{
+				CopyStmt *stmt = castNode(CopyStmt, node);
+
+				add_truncation_where_clause(state, node, stmt->whereClause);
+
+				break;
+			}
+
+		case T_InsertStmt:
+			{
+				InsertStmt *stmt = castNode(InsertStmt, node);
+
+				if (stmt->cols != NULL)
+					add_truncation(state, TRUNCATION_COLS, node, cols_len(stmt->cols));
+
+				break;
+			}
```

Comment:
We may want to add support `T_MergeStmt` here as well (which didn't exist when the truncation code was first written, so this is an existing bug, but should be easy to address here)